bitcoin/src/crypto/key.rs

```diff
@@ -863,76 +842,164 @@ impl From<TweakedKeypair> for TweakedPublicKey {
-    fn from(e: base58::Error) -> Error { Error::Base58(e) }
+impl From<base58::Error> for FromWifError {
+    fn from(e: base58::Error) -> FromWifError { Self::Base58(e) }
 }
 
-impl From<secp256k1::Error> for Error {
-    fn from(e: secp256k1::Error) -> Error { Error::Secp256k1(e) }
+impl From<secp256k1::Error> for FromWifError {
+    fn from(e: secp256k1::Error) -> FromWifError { Self::Secp256k1(e) }
 }
 
-impl From<hex::HexToArrayError> for Error {
-    fn from(e: hex::HexToArrayError) -> Self { Error::Hex(e) }
+/// Error returned while constructing public key from string.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ParsePublicKeyError {
+    /// Error originated while parsing string.
+    Encoding(FromSliceError),
+    /// Hex decoding error.
+    Hex(hex::HexToArrayError),
+    /// `PublicKey` hex should be 66 or 130 digits long.
+    InvalidHexLength(usize),
+}
+
+impl fmt::Display for ParsePublicKeyError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use ParsePublicKeyError::*;
+        match *self {
+            Encoding(ref e) => write_err!(f, "string error"; e),
+            Hex(ref e) => write_err!(f, "hex error"; e),
+            InvalidHexLength(ref got) => write!(f, "pubkey string should be 66 or 130 digits long, got: {}", got),
+        }
+    }
+}
+
+impl From<hex::HexToArrayError> for ParsePublicKeyError {
+    fn from(e: hex::HexToArrayError) -> Self { Self::Hex(e) }
+}
+
+impl From<FromSliceError> for ParsePublicKeyError {
+    fn from(e: FromSliceError) -> ParsePublicKeyError { Self::Encoding(e) }
+}
+
+impl FromStr for PublicKey {
+    type Err = ParsePublicKeyError;
+    fn from_str(s: &str) -> Result<PublicKey, ParsePublicKeyError> {
+        match s.len() {
+            66 => PublicKey::from_slice(&<[u8; 33]>::from_hex(s)?).map_err(From::from),
+            130 => PublicKey::from_slice(&<[u8; 65]>::from_hex(s)?).map_err(From::from),
+            len => Err(ParsePublicKeyError::InvalidHexLength(len)),
+        }
+    }
+}
+
+/// Error caused from CompressedPublicKey.
+#[derive(Debug)]
+pub enum ParseCompressedPublicKeyError {
+    /// Secp256k1 Error.
+    Secp256k1Error(secp256k1::Error),
+    /// hex to array conversion error.
+    HexError(hex::HexToArrayError),
+}
+
+impl fmt::Display for ParseCompressedPublicKeyError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use ParseCompressedPublicKeyError::*;
+        match *self {
+            Secp256k1Error(e) => write_err!(f, "secp256k1 error"; e),
+            HexError(e) => write_err!(f, "invalid hex"; e)
+        }
+    }
+}
+
+impl From<secp256k1::Error> for ParseCompressedPublicKeyError {
+    fn from(e: secp256k1::Error) -> ParseCompressedPublicKeyError { Self::Secp256k1Error(e) }
+}
+
+impl From<hex::HexToArrayError> for ParseCompressedPublicKeyError {
+    fn from(e: hex::HexToArrayError) -> ParseCompressedPublicKeyError { Self::HexError(e) }
+}
+
+impl FromStr for CompressedPublicKey {
```

Comment:
Why did you move this function and the `FromStr for PublicKey` function? I hope it wasn't because of my review. These functions shouldn't live near the error stuff, they live where they were, next to the `Display` impl for the type.

Reply:
Sorry, I accidentaly moved `FromStr for PublicKey and FromStr for CompressedPublicKey`.